File: programmers/42898_1.py
'''
    등굣길
    https://programmers.co.kr/learn/courses/30/lessons/42898
'''

MOD = 1000000007


def solution(m, n, puddles):
    dp = [[0] * (m + 1) for i in range(n + 1)]
    dp[1][0] = 1
    for y in range(1, n+1):
        for x in range(1, m + 1):
            if [x, y] in puddles:
                dp[y][x] = 0
            else:
                dp[y][x] = (dp[y-1][x] + dp[y][x-1]) % MOD
    return dp[n][m]


'''
    참고
    https://codedrive.tistory.com/57
'''

File: programmers/42898_1.cpp
// 등굣길
// https://programmers.co.kr/learn/courses/30/lessons/42898
#include <string>
#include <vector>
using namespace std;
typedef long long ll;
const int MAX = 100 + 1;
const ll MOD = 1000000007;
int board[MAX][MAX];
ll dp[MAX][MAX];
int solution(int m, int n, vector<vector<int>> puddles)
{
    int answer = 0;
    for (int i = 0; i < puddles.size(); i++)
    {
        board[puddles[i][1]][puddles[i][0]] = -1;
    }

    dp[1][0] = 1;
    for (int y = 1; y <= n; y++)
    {
        for (int x = 1; x <= m; x++)
        {
            if (board[y][x] == -1)
                dp[y][x] = 0;
            else
                dp[y][x] = (dp[y - 1][x] + dp[y][x - 1]) % MOD;
        }
    }

    return dp[n][m];
}